docs/architecture/adr-065-custom-event-indexing.md

```diff
@@ -0,0 +1,341 @@
+indexed by the respective type.
+
+In addition to indexing, Tendermint Core also supports the ability to query for
+both indexed transaction and block events via Tendermint's RPC layer. The ability
+to query for these indexed events facilitates a great multitude of upstream client
+and application capabilities, e.g. block explorers, IBC relayers, and auxiliary
+data availability and indexing services.
+
+Currently, Tendermint only supports indexing via a `kv` indexer, which is supported
+by an underlying embedded key/value store database. The `kv` indexer implements
+its own indexing and query mechanisms. While the former is somewhat trivial,
+providing a rich and flexible query layer is not as trivial and has caused many
+issues and UX concerns for upstream clients and applications.
+
+The fragile nature of the proprietary `kv` query engine and the potential
+performance and scaling issues that arise when a large number of consumers are
+introduced, motivate the need for a more robust and flexible indexing and query
+solution.
+
+## Alternative Approaches
+
+With regards to alternative approaches to a more robust solution, the only serious
+contender that was considered was to transition to using [SQLite](https://www.sqlite.org/index.html).
+
+While the approach would work, it locks us into a specific query language and
+storage layer, so in some ways it's only a bit better than our current approach.
+In addition, the implementation would require the introduction of CGO into the
+Tendermint Core stack, whereas right now CGO is only introduced depending on
+the database used.
+
+## Decision
+
+We will adopt a similar approach to that of the Cosmos SDK's `KVStore` state
+listening described in [ADR-038](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-038-state-listening.md).
+
+Namely, we will perform the following:
+
+- Introduce a new interface, `EventSink`, that all data sinks must implement.
+- Augment the existing `tx_index.indexer` configuration to now accept a series
+  of one or more indexer types, i.e sinks.
+- Combine the current `TxIndexer` and `BlockIndexer` into a single `KVEventSink`
+  that implements the `EventSink` interface.
+- Introduce an additional `EventSink` that is backed by [PostgreSQL](https://www.postgresql.org/).
+  - Implement the necessary schemas to support both block and transaction event
+  indexing.
+- Update `IndexerService` to use a series of `EventSinks`.
+- Proxy queries to the relevant sink's native query layer.
+- Update all relevant RPC methods.
+
+
+## Detailed Design
+
+### EventSink
+
+We introduce the `EventSink` interface type that all supported sinks must implement.
+The interface is defined as follows:
+
+```go
+type EventSink interface {
+  IndexBlockEvents(types.EventDataNewBlockHeader) error
+  IndexTxEvents(*abci.TxResult) error
+
+  SearchBlockEvents(context.Context, *query.Query) ([]int64, error)
```

Comment:
What is `query.Query` here? guessing a string but want to double check

Reply:
Correct. Resolving :)